bus/bus.go

```diff
@@ -0,0 +1,102 @@
+//
+// Copyright 2019 Insolar Technologies GbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package bus
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/message/router/middleware"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+)
+
+const ExternalMsgTopic = "ExternalMsgTopic"
+const IncomingMsgTopic = "IncomingMsgTopic"
+
+const PulseMetadataKey = "pulse"
+const TypeMetadataKey = "type"
+const ReceiverMetadataKey = "receiver"
+const SenderMetadataKey = "sender"
+
+const ReplyTypeMetadataValue = "reply"
+
+type Bus struct {
+	pub          message.Publisher
+	replies      map[string]chan *message.Message
+	repliesMutex sync.RWMutex
+	timeout      time.Duration
+}
+
+func NewBus(pub message.Publisher, timeout time.Duration) *Bus {
+	return &Bus{
+		timeout: timeout,
+		pub:     pub,
+		replies: make(map[string]chan *message.Message),
+	}
+}
+
+// Send an watermill's Message and return channel for replies.
+func (b *Bus) Send(ctx context.Context, msg *message.Message) <-chan *message.Message {
+	id := middleware.MessageCorrelationID(msg)
+	rep := make(chan *message.Message)
+	b.repliesMutex.Lock()
+	_, ok := b.replies[id]
+	if ok {
```

Comment:
This is impossible if we use `SetCorrelationID` in this function instead of `MessageCorrelationID`.